util/extsort/disk_sorter.go

```diff
@@ -131,29 +891,330 @@ func (d *DiskSorter) Sort(ctx context.Context) error {
+					removedFileNums.Store(file.fileNum, struct{}{})
+				}
+			}
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
+	// Update d.orderedFiles.
+	close(outputCh)
+	for output := range outputCh {
+		d.orderedFiles = append(d.orderedFiles, output)
+	}
+	var newOrderedFiles []*fileMetadata
+	for _, file := range d.orderedFiles {
+		if _, ok := removedFileNums.Load(file.fileNum); !ok {
+			newOrderedFiles = append(newOrderedFiles, file)
+		}
+	}
+	slices.SortFunc(newOrderedFiles, func(a, b *fileMetadata) bool {
+		return bytes.Compare(a.startKey, b.startKey) < 0
+	})
+	d.orderedFiles = newOrderedFiles
 	return nil
 }
 
-func (d *DiskSorter) doSort(_ context.Context) error {
-	iter := d.db.NewIter(nil)
-	if !iter.Last() {
-		// No keys or any error occurred.
+func (d *DiskSorter) pickCompactionFiles() []*fileMetadata {
+	type interval struct {
+		key   []byte
+		depth int
+	}
+	// intervals is a list of intervals that perfectly overlap file boundaries.
+	// For example, if we have two files [a, c] and [b, d], the intervals will be
+	// [a, b), [b, c), [c, d).
+	intervals := make([]interval, 0, len(d.orderedFiles)*2)
+	for _, file := range d.orderedFiles {
+		intervals = append(intervals, interval{
+			key:   file.startKey,
+			depth: 1,
+		})
+		intervals = append(intervals, interval{
+			key:   file.endKey,
+			depth: -1,
+		})
+	}
+	slices.SortFunc(intervals, func(a, b interval) bool {
+		return bytes.Compare(a.key, b.key) < 0
+	})
+
+	maxDepth := 0
+	for i := 1; i < len(intervals); i++ {
+		intervals[i].depth += intervals[i-1].depth
+		if intervals[i].depth > maxDepth {
+			maxDepth = intervals[i].depth
+		}
+	}
+	if maxDepth <= d.opts.CompactionThreshold {
+		return nil
+	}
+
+	var files []*fileMetadata
+	for _, file := range d.orderedFiles {
+		minIntervalIndex := sort.Search(len(intervals), func(i int) bool {
+			return bytes.Compare(intervals[i].key, file.startKey) >= 0
+		})
+		maxIntervalIndex := sort.Search(len(intervals), func(i int) bool {
+			return bytes.Compare(intervals[i].key, file.endKey) >= 0
+		})
+		for i := minIntervalIndex; i < maxIntervalIndex; i++ {
+			if intervals[i].depth >= d.opts.CompactionThreshold {
+				files = append(files, file)
+				break
+			}
+		}
+	}
+	d.opts.Logger.Info(
+		"max overlap depth exceeds the compaction threshold, pick files to compact",
+		zap.Int("maxDepth", maxDepth),
+		zap.Int("threshold", d.opts.CompactionThreshold),
+		zap.Int("fileCount", len(files)),
+	)
+	return files
+}
+
+func splitCompactionFiles(files []*fileMetadata, maxCompactionDepth int) [][]*fileMetadata {
+	// Split files into non-overlapping groups.
+	slices.SortFunc(files, func(a, b *fileMetadata) bool {
+		return bytes.Compare(a.startKey, b.startKey) < 0
+	})
+	var groups [][]*fileMetadata
+	curGroup := []*fileMetadata{files[0]}
+	maxEndKey := files[0].endKey
+	for _, file := range files[1:] {
+		if bytes.Compare(file.startKey, maxEndKey) > 0 {
```

Comment:
since maxEndKey is endKey which is exclusive, we can use >= 0?

Reply:
Yes, you are right.